vlog4j-parser/src/test/java/org/semanticweb/vlog4j/syntax/parser/RuleParserDataSourceTest.java

```diff
@@ -0,0 +1,144 @@
+		RdfFileDataSource rdfds = new RdfFileDataSource(new File(EXAMPLE_RDF_FILE_PATH));
+		assertEquals(rdfds, RuleParser.parseDataSourceDeclaration(input));
+	}
+
+	@Test(expected = ParsingException.class)
+	public void testRdfSourceInvalidArity() throws ParsingException, IOException {
+		String input = "@source p[2] : load-rdf(\"" + EXAMPLE_RDF_FILE_PATH + "\") .";
+		RuleParser.parseDataSourceDeclaration(input);
+	}
+
+	@Test
+	public void testSparqlSource() throws ParsingException, MalformedURLException {
+		String input = "@source p[2] : sparql(<" + WIKIDATA_SPARQL_ENDPOINT_URI + ">,\"disease, doid\",\"?disease wdt:P699 ?doid .\") .";
+		SparqlQueryResultDataSource sparqlds = new SparqlQueryResultDataSource(
+				new URL(WIKIDATA_SPARQL_ENDPOINT_URI), "disease, doid", "?disease wdt:P699 ?doid .");
+		assertEquals(sparqlds, RuleParser.parseDataSourceDeclaration(input));
+	}
+
+	@Test(expected = ParsingException.class)
+	public void testSparqlSourceMalformedUrl() throws ParsingException, MalformedURLException {
+		String input = "@source p[2] : sparql(<not a URL>,\"disease, doid\",\"?disease wdt:P699 ?doid .\") .";
+		RuleParser.parseDataSourceDeclaration(input);
+	}
+
+	@Test(expected = ParsingException.class)
+	public void testUnknownDataSource() throws ParsingException {
+		String input = "@source p[2] : unknown-data-source(\"hello, world\") .";
+		RuleParser.parseDataSourceDeclaration(input);
+	}
+
+	@Test
+	public void testCustomDataSource() throws ParsingException {
+		CsvFileDataSource source = mock(CsvFileDataSource.class);
+		DataSourceDeclarationHandler handler = mock(DataSourceDeclarationHandler.class);
+		ParserConfiguration parserConfiguration = new ParserConfiguration();
+		parserConfiguration.registerDataSource("mock-source", handler);
+		doReturn(source).when(handler).handleDeclaration(ArgumentMatchers.<List<String>>any(),
+				ArgumentMatchers.<SubParserFactory>any());
+
+		String input = "@source p[2] : mock-source(\"hello\", \"world\") .";
+		List<String> expectedArguments = Arrays.asList("hello", "world");
+		RuleParser.parseDataSourceDeclaration(input, parserConfiguration);
+
+		verify(handler).handleDeclaration(eq(expectedArguments), ArgumentMatchers.<SubParserFactory>any());
+	}
+
+	@Test
+	public void sparqlDataSourceDeclarationToStringParsingTest() throws ParsingException, IOException {
+		KnowledgeBase kb = new KnowledgeBase();
+		Predicate predicate1 = Expressions.makePredicate("p", 3);
+		SparqlQueryResultDataSource dataSource = new SparqlQueryResultDataSource(new URL(WIKIDATA_SPARQL_ENDPOINT_URI),
+				"var", "?var wdt:P31 wd:Q5 .");
+		DataSourceDeclaration dataSourceDeclaration1 = new DataSourceDeclarationImpl(predicate1, dataSource);
+		RuleParser.parseInto(kb, dataSourceDeclaration1.toString());
+		assertEquals(dataSourceDeclaration1, kb.getDataSourceDeclarations().get(0));
+	}
+
+	@Test
+	public void rdfDataSourceDeclarationToStringParsingTest() throws ParsingException, IOException {
```

Comment:
Fails on windows: 
`Caused by: org.semanticweb.vlog4j.parser.javacc.TokenMgrError: Lexical error at line 1, column 29.  Encountered: "m" (109), after : "\"src\\"`

Until issue (https://github.com/knowsys/vlog4j/issues/140) is fixed, just test a simple path constructed using File.separator

Reply:
No, this looks like a bug in the parser, it should accept strings with backslashes in them.

Reply:
Okay, no, it's actually a bug in the Serializer that doesn't properly escape the file names in data source declarations.